libpromises/cf3parse.y

```diff
@@ -573,6 +574,88 @@ constraint:            constraint_id                        /* BUNDLE ONLY */
+                                           else if (P.rval.type == RVAL_TYPE_SCALAR &&
+                                                    (strcmp(P.lval, "data") == 0 || strcmp(P.lval, "template_data") == 0))
+                                           {
+                                               JsonElement *json = NULL;
+                                               JsonParseError res;
+                                               bool json_parse_attempted = false;
+                                               Buffer *copy = BufferNewFrom(P.rval.item, strlen(P.rval.item));
+
+                                               const char* fname = NULL;
+                                               int offset = 0;
+                                               if (strlen(P.rval.item) > 3 && strncmp("---", P.rval.item, 3) == 0)
+                                               {
+                                                   fname = "parseyaml";
+                                                   offset = 0;
+
+                                                   // look for unexpanded variables
+                                                   if (NULL == strstr(P.rval.item, "$(") && NULL == strstr(P.rval.item, "${"))
+                                                   {
+                                                       const char *copy_data = BufferData(copy);
+                                                       res = JsonParseYamlString(&copy_data, &json);
+                                                       json_parse_attempted = true;
+                                                   }
+                                               }
+                                               else
+                                               {
+                                                   fname = "parsejson";
+                                                   offset = 0;
+                                                   // look for unexpanded variables
+                                                   if (NULL == strstr(P.rval.item, "$(") && NULL == strstr(P.rval.item, "${"))
+                                                   {
+                                                       const char *copy_data = BufferData(copy);
+                                                       res = JsonParse(&copy_data, &json);
+                                                       json_parse_attempted = true;
+                                                   }
+                                               }
+
+                                               BufferDestroy(copy);
+
+                                               if (json_parse_attempted && res != JSON_PARSE_OK)
+                                               {
+                                                   // no error, we'll try with the fname fncall
```

Comment:
Slightly confusing comment, maybe instead: ``Parsing failed, insert fncall so it can be retried during evaluation``.